### PR TITLE
Fixed #34686 -- Addded support for GEOS 3.12.

### PR DESCRIPTION
**Pull request by olibook**
Fix for https://code.djangoproject.com/ticket/34686

### CODE REVIEW

File: tests/gis_tests/geos_tests/test_io.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def test01_wktreader(self):
 
     def test02_wktwriter(self):
         # Creating a WKTWriter instance, testing its ptr property.
-        wkt_w = WKTWriter()
+        wkt_w = WKTWriter(trim=False)
```
Comment (felixxm) [inline tests/gis_tests/geos_tests/test_io.py:38]: This seems unnecessary as `trim=False` is the default value.

File: django/contrib/gis/geos/prototypes/io.py [hunk 2]
```diff
@@ -211,7 +219,7 @@ def outdim(self, new_dim):
 
     @property
     def trim(self):
-        return self._trim
+        return default_trim_value() if self._trim is None else self._trim
```
Comment (felixxm) [inline django/contrib/gis/geos/prototypes/io.py:222]: I have some doubts about this change. First of all, calculating `default_trim_value()` every time looks unnecessary, can we move this to a lazy module constant or something similar? Secondly, `default_trim_value()` is not taken into account in the setter, this shouldn't work :thinking: 

I'd start from removing a redundant logic in a separate PR, e.g.
```diff
diff --git a/django/contrib/gis/geos/prototypes/io.py b/django/contrib/gis/geos/prototypes/io.py
index f291ec08b6..20ef8b37ce 100644
--- a/django/contrib/gis/geos/prototypes/io.py
+++ b/django/contrib/gis/geos/prototypes/io.py
@@ -189,8 +189,7 @@ class WKTWriter(IOBase):
 
     def __init__(self, dim=2, trim=False, precision=None):
         super().__init__()
-        if bool(trim) != self._trim:
-            self.trim = trim
+        self.trim = trim
         if precision is not None:
             self.precision = precision
         self.outdim = dim
```
What do you think?

File: django/contrib/gis/geos/prototypes/io.py [hunk 3]
```diff
@@ -178,18 +178,26 @@ def read(self, wkb):
             raise TypeError
 
 
+def default_trim_value():
+    """
+    Geos changed the default value in 3.12.0: https://github.com/libgeos/geos/pull/915
```
Comment (felixxm) [inline django/contrib/gis/geos/prototypes/io.py:183]: `Geos` -> `GEOS`. Also, remove a link to the PR.

File: django/contrib/gis/geos/prototypes/io.py [hunk 4]
```diff
@@ -178,18 +178,26 @@ def read(self, wkb):
...
+    can be replaced by a static True when 3.12.0 becomes the minimum supported version
+    """
+    return geos_version_tuple() >= (3, 12, 0)
```
Comment (felixxm) [inline django/contrib/gis/geos/prototypes/io.py:186]:
```suggestion
    return geos_version_tuple() >= (3, 12)
```

**Review by felixxm — COMMENTED** [review]
@olibook Thanks :+1: I left comments.

Please add a release note (`5.0.txt`), e.g.
```
* Added support for GEOS 3.12.
```

File: django/contrib/gis/geos/prototypes/io.py (hunk 2 above)
Comment (olibook) [reply django/contrib/gis/geos/prototypes/io.py:222]: Hi Felix,

A lazy module constant definitely looks like a good idea. 

I'll simplify the trim getter setter changes

File: tests/gis_tests/geos_tests/test_io.py (hunk 1 above)
Comment (olibook) [reply tests/gis_tests/geos_tests/test_io.py]: fixed :-)

File: django/contrib/gis/geos/prototypes/io.py (hunk 4 above)
Comment (olibook) [reply django/contrib/gis/geos/prototypes/io.py]: fixed

File: django/contrib/gis/geos/prototypes/io.py [hunk 5]
```diff
@@ -178,19 +179,28 @@ def read(self, wkb):
...
-
-    _trim = False
     _precision = None
 
     def __init__(self, dim=2, trim=False, precision=None):
         super().__init__()
-        if bool(trim) != self._trim:
-            self.trim = trim
+        self._trim = bool(DEFAULT_TRIM_VALUE)
+        self.trim = trim
```
Comment (felixxm) [inline django/contrib/gis/geos/prototypes/io.py:203]: I moved this to a separate PR, check out #17063.

Comment (olibook) [reply django/contrib/gis/geos/prototypes/io.py:203]: merged

Comment (olibook) [reply django/contrib/gis/geos/prototypes/io.py:203]: we still need to set the default trim value in an object instance variable so that we can detect the `trim` parameter value is different from what is stored in the underlying GEOS object, so that the function `wkt_writer_set_trim` is called appropriately depending on GEOS version, i mean:
- if geos < 3.12 and trim parameter is False, do not call wkt_writer_set_trim
- if geos >= 3.12 and trim parameter is True, do not call wkt_writer_set_trim

**Comment by smithdc1** [comment]
Hi @olibook 

Thank you for picking this up. 👍 

Running with the spatialite backend I see two test failures with this ticket and GEOS 3.12.

```
======================================================================
ERROR: test_make_valid (gis_tests.geoapp.test_functions.GISFunctionsTests.test_make_valid)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
    yield
  File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 623, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/workspaces/django/django/test/testcases.py", line 1428, in skip_wrapper
    return test_func(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/workspaces/django/tests/gis_tests/geoapp/test_functions.py", line 489, in test_make_valid
    self.assertIs(invalid.repaired.valid, True)
    ^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'valid'

======================================================================
ERROR: test_make_valid_multipolygon (gis_tests.geoapp.test_functions.GISFunctionsTests.test_make_valid_multipolygon)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
    yield
  File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 623, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/workspaces/django/django/test/testcases.py", line 1428, in skip_wrapper
    return test_func(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/workspaces/django/tests/gis_tests/geoapp/test_functions.py", line 509, in test_make_valid_multipolygon
    self.assertIs(invalid.repaired.valid, True)
    ^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'valid'
```

**Comment by olibook** [comment]
> Hi @olibook
> 
> Thank you for picking this up. 👍
> 
> Running with the spatialite backend I see two test failures with this ticket and GEOS 3.12.
> 
> ```
> ======================================================================
> ERROR: test_make_valid (gis_tests.geoapp.test_functions.GISFunctionsTests.test_make_valid)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
>     yield
>   File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 623, in run
>     self._callTestMethod(testMethod)
>     ^^^^^^^^^^^^^^^^^
>   File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
>     if method() is not None:
>     ^^^^^^^^^^^^^^^^^
>   File "/workspaces/django/django/test/testcases.py", line 1428, in skip_wrapper
>     return test_func(*args, **kwargs)
>     ^^^^^^^^^^^^^^^^^
>   File "/workspaces/django/tests/gis_tests/geoapp/test_functions.py", line 489, in test_make_valid
>     self.assertIs(invalid.repaired.valid, True)
>     ^^^^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'valid'
> 
> ======================================================================
> ERROR: test_make_valid_multipolygon (gis_tests.geoapp.test_functions.GISFunctionsTests.test_make_valid_multipolygon)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
>     yield
>   File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 623, in run
>     self._callTestMethod(testMethod)
>     ^^^^^^^^^^^^^^^^^
>   File "/opt/conda/envs/django/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
>     if method() is not None:
>     ^^^^^^^^^^^^^^^^^
>   File "/workspaces/django/django/test/testcases.py", line 1428, in skip_wrapper
>     return test_func(*args, **kwargs)
>     ^^^^^^^^^^^^^^^^^
>   File "/workspaces/django/tests/gis_tests/geoapp/test_functions.py", line 509, in test_make_valid_multipolygon
>     self.assertIs(invalid.repaired.valid, True)
>     ^^^^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'valid'
> ```

Will have a look at these failures. Thanks

**Comment by olibook** [comment]
Hi @smithdc1 , could you elaborate on your setup ? 

tests are ok on my setup
-  linux ubuntu lunar, on amd64
- libsqlite3-mod-spatialite 5.0.1-3 installed with apt
- python 3.11.2 installed with apt
- gdal 3.6.2 installed with apt
- geos 3.12.0 built from sources

geos version is reported as 3.12.0 when using `geos_version_tuple`

**Comment by olibook** [comment]
Hi @smithdc1 thank you for your input. I reproduced the test failure, digged into it and it seems to be an issue with spatialite and GEOS 3.12.0, not with django itself.

it works when GEOS is pinned to 3.11.2

```
$ conda create -n django_geos311 -c conda-forge 'geos==3.11.2' gdal proj libspatialite sqlite
$ conda activate django_geos311
$ sqlite3
sqlite> SELECT load_extension('mod_spatialite');   
select AsWKT(MakeValid(GeomFromText('POLYGON ((0 0, 0 1, 1 1, 1 0, 1 1, 1 0, 0 0))')));
POLYGON((0 0,0 1,1 1,1 0,0 0))
```

Same SQL Query returns an empty geometry (hence test failures) when the conda env has GEOS pinned to 3.12.0

```
$ conda create -n django_geos312 -c conda-forge 'geos==3.12.0' gdal proj libspatialite sqlite
$ conda activate django_geos312
$ sqlite3
sqlite> SELECT load_extension('mod_spatialite');   
select AsWKT(MakeValid(GeomFromText('POLYGON ((0 0, 0 1, 1 1, 1 0, 1 1, 1 0, 0 0))')));
GEOMETRYCOLLECTION()
```

I went a bit further and all tests are passing when :
- django uses GEOS 3.12.0
- spatialite uses GEOS 3.11.2

I'll get in touch with spatialite community to get some insights

**Comment by olibook** [comment]
Update on this topic:

- `test_make_valid_*` related tests are ok when running postgis / postgres 15 / geos 3.12.0
- I'm digging deeper and issue may be in librttopo in the [RTGEOM_GEOS_makeValidPolygon](https://github.com/CGX-GROUP/librttopo/blob/master/src/rtgeom_geos_clean.c#L380) function.

next step would be to build spatialite with librttopo with RTGEOM_PROFILE_MAKEVALID=1, and compare traces between geos 3.11 and geos 3.12. 

What do you think ?

**Comment by olibook** [comment]
Update on this topic:

The spatiality community [acknowledged the issue](https://groups.google.com/g/spatialite-users/c/6LPgCHn3Bbs/m/fX8kUZp8AwAJ?pli=1) yesterday and is preparing a patch for GEOS 3.12.0

**Comment by olibook** [comment]
Update on this topic:

spatialite 5.1.0 will support GEOS 3.12. There is a release candidate available.

- https://groups.google.com/g/spatialite-users/c/hAJ2SgitN4M/m/VnuR2l1sBAAJ

Will update the PR upon spatialite public release.

**Comment by felixxm** [comment]
@olibook Thanks for all your efforts :1st_place_medal: 

> Will update the PR upon spatialite public release.

Do we need to update anything after the SpatiaLite 5.1final release? I think we can move it forward without it, as the lack of SpatiaLite support for GEOS 3.12 is not something Django should be worry about.

File: django/contrib/gis/geos/prototypes/io.py [hunk 6]
```diff
@@ -178,17 +179,27 @@ def read(self, wkb):
...
     _constructor = wkt_writer_create
     ptr_type = WKT_WRITE_PTR
     destructor = wkt_writer_destroy
-
-    _trim = False
     _precision = None
 
     def __init__(self, dim=2, trim=False, precision=None):
         super().__init__()
+        self._trim = bool(DEFAULT_TRIM_VALUE)
```
Comment (felixxm) [inline django/contrib/gis/geos/prototypes/io.py:202]: `bool()` is unnecessary
```suggestion
        self._trim = DEFAULT_TRIM_VALUE
```

**Comment by olibook** [comment]
Spatialite 5.1.0 RC1 has a regression on this test:

```
FAIL: test_asgml (gis_tests.geoapp.test_functions.GISFunctionsTests.test_asgml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
    yield
  File "/usr/lib/python3.11/unittest/case.py", line 623, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/home/olivier.linux/django/django/test/testcases.py", line 1425, in skip_wrapper
    return test_func(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/olivier.linux/django/tests/gis_tests/geoapp/test_functions.py", line 147, in test_asgml
    self.assertTrue(gml_regex.match(ptown.gml))
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/case.py", line 715, in assertTrue
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^^^
AssertionError: None is not true

----------------------------------------------------------------------
Ran 557 tests in 0.963s
```

for the record: `ptown.gml` contains

```
ipdb> ptown.gml
'<gml:Point srsName="urn:ogc:def:crs:EPSG:4326"><gml:coordinates>-104.609252,38.255001</gml:coordinates></gml:Point>'
```

I think this would be a separate ticket to assess compatibility with spatialite 5.1

File: tests/gis_tests/geoapp/models.py [hunk 7]
```diff
@@ -95,3 +95,7 @@ class ManyPointModel(NamedModel):
     point1 = models.PointField()
     point2 = models.PointField()
     point3 = models.PointField(srid=3857)
+
+
+class ThreeDimensionalFeature(NamedModel):
+    geom = models.GeometryField(dim=3)
```
Comment (felixxm) [inline tests/gis_tests/geoapp/models.py]: I added `required_db_features = {"supports_3d_storage"}`, check out fed257ddff449e1bf1ce06d9a7fddb148290a6ac.

**Review by felixxm — COMMENTED** [review]
@olibook Thanks :+1: I pushed final edits.

@timgraham Can you confirm that it works on CockroachDB?

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/gis_tests/geoapp/models.py (hunk 7 above)
Comment (timgraham) [reply tests/gis_tests/geoapp/models.py]: All tests pass on CockroachDB. Since the commit you mentioned, CockroachDB added support for 3D storage so the feature flag isn't used there any longer.

Comment (felixxm) [reply tests/gis_tests/geoapp/models.py]: Many thanks for checking :+1:
